File: Scales/app/src/main/java/com/example/scales/ScalesActivity.java
package com.example.scales;

import android.bluetooth.*;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.widget.TextView;

import java.nio.charset.Charset;
import java.util.Objects;
import java.util.UUID;

public class ScalesActivity extends AppCompatActivity {

    private static final String TAG = "Iksde";

    private boolean mConnected;
    private BluetoothAdapter mBluetoothAdapter;
    private BluetoothGatt mGatt;
    BluetoothDevice device;
    int retries = 10;
    TextView read;


    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_scales);

        BluetoothManager bluetoothManager = (BluetoothManager) getSystemService(BLUETOOTH_SERVICE);
        mBluetoothAdapter = bluetoothManager.getAdapter();

        read = findViewById(R.id.weight_value);
        device = mBluetoothAdapter.getBondedDevices().stream().filter(mBluetoothAdapter -> mBluetoothAdapter.getName().startsWith("BL weight")).findAny().orElse(null);
        new Thread(() -> connectDevice(Objects.requireNonNull(device))).start();
    }


    private void connectDevice(BluetoothDevice device) {
        log("Connecting to " + device.getAddress());
        GattClientCallback gattClientCallback = new GattClientCallback();
        mGatt = device.connectGatt(this, false, gattClientCallback, 2);
    }

    // Messaging

    private void sendMessage(String msg) {
        if (!mConnected) {
            return;
        }

        BluetoothGattCharacteristic characteristic =  mGatt.getServices().get(2)
                .getCharacteristics()
                .stream().filter(bluetoothGattCharacteristic -> bluetoothGattCharacteristic.getUuid().equals(UUID.fromString("6E400002-B5A3-F393-E0A9-E50E24DCCA9E")))
                .findAny().orElse(null);


        byte[] message = msg.getBytes(Charset.forName("UTF-8"));

        if (characteristic != null) {
            characteristic.setValue(message);
            boolean success = mGatt.writeCharacteristic(characteristic);
            if (success) {
                log("Success");
            } else {
                log("Przypeks");
            }
        }

    }

    public void log(String msg) {
        Log.d(TAG, msg);
    }


    public void setConnected(boolean connected) {
        mConnected = connected;
    }

    public void disconnectGattServer() {
        mConnected = false;
        if (mGatt != null) {
            mGatt.disconnect();
            mGatt.close();
        }
        if (retries > 0) {
            retries--;
            connectDevice(device);
        }
    }

    private class GattClientCallback extends BluetoothGattCallback {

        @Override
        public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState) {
            super.onConnectionStateChange(gatt, status, newState);
            log("onConnectionStateChange newState: " + newState);

            if (status == BluetoothGatt.GATT_FAILURE) {
                log("Connection Gatt failure status " + status);
                disconnectGattServer();
                return;
            }

            if (newState == BluetoothProfile.STATE_CONNECTED) {
                log("Connected to device " + gatt.getDevice().getAddress());
                setConnected(true);
                gatt.discoverServices();
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                log("Disconnected from device");
                disconnectGattServer();
            }
        }

        @Override
        public void onServicesDiscovered(BluetoothGatt gatt, int status) {
            super.onServicesDiscovered(gatt, status);

            if (status != BluetoothGatt.GATT_SUCCESS) {
                log("Device service discovery unsuccessful, status " + status);
                return;
            }

            mGatt.getServices().forEach(bluetoothGattService -> {
                log(bluetoothGattService.getUuid().toString());
            });

            BluetoothGattCharacteristic characteristic =  mGatt.getServices().get(2)
                    .getCharacteristics()
                    .stream().filter(bluetoothGattCharacteristic -> {
                        log(bluetoothGattCharacteristic.getUuid().toString());
                        return bluetoothGattCharacteristic.getUuid().equals(UUID.fromString("6E400003-B5A3-F393-E0A9-E50E24DCCA9E"));
                    })
                    .findAny().orElse(null);
             enableCharacteristicNotification(gatt, characteristic);
        }

        @Override
        public void onCharacteristicWrite(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic, int status) {
            super.onCharacteristicWrite(gatt, characteristic, status);
            if (status == BluetoothGatt.GATT_SUCCESS) {
                log("Characteristic written successfully");
            } else {
                log("Characteristic write unsuccessful, status: " + status);
                disconnectGattServer();
            }
        }

        @Override
        public void onCharacteristicRead(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic, int status) {
            super.onCharacteristicRead(gatt, characteristic, status);
            if (status == BluetoothGatt.GATT_SUCCESS) {
                log("Characteristic read successfully");
                readCharacteristic(characteristic);
            } else {
                log("Characteristic read unsuccessful, status: " + status);
            }
        }

        @Override
        public void onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
            super.onCharacteristicChanged(gatt, characteristic);
            log("Characteristic changed, " + characteristic.getUuid().toString());
            readCharacteristic(characteristic);
        }

        private void enableCharacteristicNotification(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
            boolean characteristicWriteSuccess = gatt.setCharacteristicNotification(characteristic, true);
            if (characteristicWriteSuccess) {
                log("Characteristic notification set successfully for " + characteristic.getUuid().toString());
            } else {
                log("Characteristic notification set failure for " + characteristic.getUuid().toString());
            }
        }

        private void readCharacteristic(BluetoothGattCharacteristic characteristic) {
            byte[] messageBytes = characteristic.getValue();
            String message = new String(messageBytes);
            log("Received message: " + message);
            if (message.length() > 3) {
                read.setText(message);
            }
        }
    }
}

File: Scales/app/src/main/java/com/example/scales/ScalesActivity.kt
package com.example.scales

import android.app.Activity
import android.bluetooth.*
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.nio.charset.Charset
import java.util.*


class ScalesActivity : Activity() {
    lateinit var myLabel: TextView
    lateinit var myTextbox: EditText
    var mBluetoothAdapter: BluetoothAdapter? = null
    lateinit var mmSocket: BluetoothSocket
    lateinit var mmDevice: BluetoothDevice
    lateinit var mmOutputStream: OutputStream
    lateinit var mmInputStream: InputStream
    lateinit var workerThread: Thread
    lateinit var readBuffer: ByteArray
    var readBufferPosition: Int = 0
    var counter: Int = 0
    @Volatile
    var stopWorker: Boolean = false

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_scales)

        val openButton = findViewById(R.id.open) as Button
        val sendButton = findViewById(R.id.send) as Button
        val closeButton = findViewById(R.id.close) as Button
        myLabel = findViewById(R.id.label) as TextView
        myTextbox = findViewById(R.id.entry) as EditText

        //Open Button
        openButton.setOnClickListener {
                findBT()
                openBT()
        }

        //Send Button
        sendButton.setOnClickListener {
            try {
                sendData()
            } catch (ex: IOException) {
            }
        }

        //Close button
        closeButton.setOnClickListener {
            try {
                closeBT()
            } catch (ex: IOException) {
            }
        }
    }

    fun findBT() {
        mBluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
        if (mBluetoothAdapter == null) {
            myLabel.text = "No bluetooth adapter available"
        }

        if (!mBluetoothAdapter!!.isEnabled) {
            val enableBluetooth = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
            startActivityForResult(enableBluetooth, 0)
        }

        val pairedDevices = mBluetoothAdapter!!.bondedDevices
        if (pairedDevices.size > 0) {
            for (device in pairedDevices) {
                if (device.name == "BL weight") {
                    mmDevice = device
                    break
                }
            }
        }
        myLabel.text = "Bluetooth Device Found"
    }

    @Throws(IOException::class)
    fun openBT() {


        val gattCallback = BluetoothGattCallback() {

            }
        }
        var gatt: BluetoothGatt = mmDevice.connectGatt(this, true, gattCallback)


        mmSocket.connect()

        mmOutputStream = mmSocket.outputStream

        Log.i("KURWA", mmSocket.isConnected.toString())
        beginListenForData()

        myLabel.text = "Bluetooth Opened"
    }

    internal fun beginListenForData() {
        val handler = Handler()
        val delimiter: Byte = 10 //This is the ASCII code for a newline character

        stopWorker = false
        readBufferPosition = 0
        readBuffer = ByteArray(1024)
        workerThread = Thread(Runnable {
            while (!Thread.currentThread().isInterrupted && !stopWorker) {
                try {
                    val bytesAvailable = mmInputStream.available()
                    if (bytesAvailable > 0) {
                        val packetBytes = ByteArray(bytesAvailable)
                        mmInputStream.read(packetBytes)
                        for (i in 0 until bytesAvailable) {
                            val b = packetBytes[i]
                            if (b == delimiter) {
                                val encodedBytes = ByteArray(readBufferPosition)
                                System.arraycopy(readBuffer, 0, encodedBytes, 0, encodedBytes.size)
                                val data = String(encodedBytes, Charset.forName("US-ASCII"))
                                readBufferPosition = 0

                                handler.post { myLabel.text = data }
                            } else {
                                readBuffer[readBufferPosition++] = b
                            }
                        }
                    }
                } catch (ex: IOException) {
                    stopWorker = true
                }

            }
        })

        workerThread.start()
    }

    @Throws(IOException::class)
    internal fun sendData() {
        var msg = myTextbox.text.toString()
        msg += "\n"
        mmOutputStream.write(msg.toByteArray())
        myLabel.text = "Data Sent"
    }

    @Throws(IOException::class)
    internal fun closeBT() {
        stopWorker = true
        mmOutputStream.close()
        mmInputStream.close()
        mmSocket.close()
        myLabel.text = "Bluetooth Closed"
    }
}